internal/pkg/mining/worker_test.go

```diff
@@ -31,14 +31,130 @@ import (
-	mtg.ticketGen = true
+	mtg.fn(ticket)
 	return consensus.MakeFakeTicketForTest(), nil
 }
 
+type mockElectionMachine struct {
+	fn func(block.Ticket)
+}
+
+func newMockElectionMachine(f func(block.Ticket)) *mockElectionMachine {
+	return &mockElectionMachine{fn: f}
+}
+
+func (mem *mockElectionMachine) RunElection(ticket block.Ticket, candidateAddr address.Address, signer types.Signer, nullCount uint64) (block.VRFPi, error) {
+	mem.fn(ticket)
+	return consensus.MakeFakeElectionProofForTest(), nil
+}
+func (fem *mockElectionMachine) IsElectionWinner(ctx context.Context, ptv consensus.PowerTableView, ticket block.Ticket, nullCount uint64, electionProof block.VRFPi, signerAddr, minerAddr address.Address) (bool, error) {
+	return true, nil
+}
+
+func TestLookbackElection(t *testing.T) {
+	tf.UnitTest(t)
+
+	mockSignerVal, blockSignerAddr := setupSigner()
+	mockSigner := &mockSignerVal
+
+	builder := chain.NewBuilder(t, address.Undef)
+	lookback := consensus.ElectionLookback
+	head := builder.AppendManyOn(lookback-1, builder.NewGenesis())
+	ancestors := builder.RequireTipSets(head.Key(), lookback)
+
+	st, pool, addrs, bs := sharedSetup(t, mockSignerVal)
+	getStateTree := func(c context.Context, ts block.TipSet) (state.Tree, error) {
+		return st, nil
+	}
+	getAncestors := func(ctx context.Context, ts block.TipSet, newBlockHeight *types.BlockHeight) ([]block.TipSet, error) {
+		return ancestors, nil
+	}
+
+	minerAddr := addrs[3]      // addr4 in sharedSetup
+	minerOwnerAddr := addrs[4] // addr5 in sharedSetup
+
+	messages := chain.NewMessageStore(bs)
+
+	t.Run("Election sees ticket lookback ancestors back", func(t *testing.T) {
+		electionTicket, err := ancestors[lookback-1].MinTicket()
+		require.NoError(t, err)
+		mem := newMockElectionMachine(func(ticket block.Ticket) {
+			assert.Equal(t, electionTicket, ticket)
+		})
+
+		ctx, cancel := context.WithCancel(context.Background())
+		outCh := make(chan mining.Output)
+		worker := mining.NewDefaultWorker(mining.WorkerParameters{
+			API: th.NewDefaultFakeWorkerPorcelainAPI(blockSignerAddr),
+
+			MinerAddr:      minerAddr,
+			MinerOwnerAddr: minerOwnerAddr,
+			WorkerSigner:   mockSigner,
+
+			GetStateTree: getStateTree,
+			GetWeight:    getWeightTest,
+			GetAncestors: getAncestors,
+			Election:     mem,
+			TicketGen:    &consensus.FakeTicketMachine{},
+
+			MessageSource: pool,
+			Processor:     th.NewFakeProcessor(),
+			Blockstore:    bs,
+			MessageStore:  messages,
+			Clock:         clock.NewSystemClock(),
+		})
+
+		go worker.Mine(ctx, head, 0, outCh)
+		r := <-outCh
+		assert.NoError(t, r.Err)
+		cancel()
+	})
+
+	t.Run("Ticket gensees ticket 1 ancestor back", func(t *testing.T) {
+		genTicket, err := ancestors[0].MinTicket()
+		require.NoError(t, err)
+		mtm := newMockTicketGen(func(ticket block.Ticket) {
+			assert.Equal(t, genTicket, ticket)
+		})
+
+		ctx, cancel := context.WithCancel(context.Background())
```

Comment:
This context (and the one in the previous test) should also have a timeout to keep this test from hanging.

Reply:
Only if it's very long, else the test will flake on heavily-loaded systems. Test code can't make good judgements about the performance of the system it's running on. A timeout can also be provided from the outside.

Reply:
I opted not to add the timeout for @anorth's reason above.  We can revisit in an issue if you'd like to file one @acruikshank. We should probably clean up the entire channel-based interface and return structure of the worker in one go and this could be a subset of that technical debt work.